psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -888,24 +976,432 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+                outfile->Printf("    Shell: %4d, Size: %4d, Offset: %4d, Offset2: %4d\n", P, size, off,
+                                off2);
+            }
+        }
+        outfile->Printf("\n");
+    }
+
+    // => Store significant atom pairs for parallelization <= //
+    std::vector<std::pair<int, int>> atom_pairs;
+    for (size_t Patom = 0; Patom < natom; Patom++) {
+        for (size_t Qatom = 0; Qatom <= Patom; Qatom++) {
+            bool found = false;
+            for (int P = shell_endpoints_for_atom[Patom]; P < shell_endpoints_for_atom[Patom + 1]; P++) {
+                for (int Q = shell_endpoints_for_atom[Qatom]; Q < shell_endpoints_for_atom[Qatom + 1]; Q++) {
+                    if (ints[0]->shell_pair_significant(P, Q)) {
+                        found = true;
+                        atom_pairs.emplace_back(Patom, Qatom);
+                        break;
+                    }
+                }
+                if (found) break;
+            }
+        }
+    }
+
+    // => Store shell pairs that are significant by Schwarz estimate <= //
+    // A comparator used for sorting integral screening values
+    auto screen_compare = [](const std::pair<int, double> &a, 
+                                    const std::pair<int, double> &b) { return a.second > b.second; };
+
+    std::vector<std::vector<int>> significant_bras(nshell);
+    double max_integral = ints[0]->max_integral();
+
+    for (size_t P = 0; P < nshell; P++) {
+        std::vector<std::pair<int, double>> PQ_shell_values;
+        for (size_t Q = 0; Q < nshell; Q++) {
+            double pq_pq = std::sqrt(ints[0]->shell_ceiling2(P, Q, P, Q));
+            double schwarz_value = std::sqrt(pq_pq * max_integral);
+            if (schwarz_value >= cutoff_) {
+                PQ_shell_values.emplace_back(Q, schwarz_value);
+            }
+        }
+        std::sort(PQ_shell_values.begin(), PQ_shell_values.end(), screen_compare);
+
+        for (const auto& value : PQ_shell_values) {
+            significant_bras[P].push_back(value.first);
+        }
+    }
+
+    // => Calculate Shell Ceilings (To find significant bra-ket pairs) <=
+    // sqrt(Umax|Umax) in Ochsenfeld Eq. 3
+    std::vector<double> shell_ceilings(nshell, 0.0);
+    for (int P = 0; P < nshell; P++) {
+        for (int Q = 0; Q <= P; Q++) {
+            double val = std::sqrt(ints[0]->shell_ceiling2(P, Q, P, Q));
+            shell_ceilings[P] = std::max(shell_ceilings[P], val);
+            shell_ceilings[Q] = std::max(shell_ceilings[Q], val);
+        }
+    }
+
+    // => "Pre-ordering and Pre-selection to find significant elements in Puv" <= //
+
+    std::vector<std::vector<int>> significant_kets(nshell);
+
+    // => Defined in Ochsenfeld Eq. 3 <= //
+    // For shells P and R, If |Dpr| * sqrt(Pmax|Pmax) * sqrt(Rmax|Rmax) [screening value] >= linK_ints_cutoff_,
+    // Then shell R is added to the ket list of shell P (and sorted by the screening value)
```

Comment:
```suggestion
    // ==> Use shell ceilings to compute significant ket-shells for each bra-shell <==
```
I don't find any of the old comments helpful: I'd be better off reading the code for that, or looking at Figure 1.